fixedexercise3.1/NumberToLetterConverter.java

```diff
@@ -0,0 +1,141 @@
+/**
+ *
+ * @author roberta
+ */
+public abstract class NumberToLetterConverter {
+
+    private static final String[] UNITS = {"", "One ", "Two ", "Three ",
+        "Four ", "Five ", "Six ", "Seven ", "Eight ", "Nine ", "Ten ",
+        "Eleven ", "Twelve ", "Thirteen ", "Fourteen ", "Fifteen ", "Sixteen",
+        "Seventeen", "Eighteen", "Nineteen", "Twenty"};
+    private static final String[] TENS = {"Twenty", "Thirty ", "Forty ",
+        "Fifty ", "Sixty ", "Seventy ", "Eighty ", "Ninety ",
+        "One Hundred "};
+    private static final String[] HUNDREDS = {"One Hundred ", "Two Hundred ",
+        "Three Hundred ", "Four Hundred ", "Five Hundred ", "Six Hundred ",
+        "Seven Hundred ", "Eight Hundred ", "Nine Hundred  "};
+
+    /**
+     * El número es válido si está entre 0 y 999.999.999
+     */
+    public static String convertNumberToLetter(double number)
+            throws NumberFormatException {
+        String converted = new String();
+
+        // validate if a legal number
+        if (number < 0) {
+            throw new IllegalArgumentException("Can't convert negative numbers");
+        }
+        double doubleNumber = Math.rint(number * 100) / 100;
+        if (doubleNumber > 999999999) {
+            throw new NumberFormatException(
+                    "The number is greater than 999.999.999, "
+                    + "isn't possible to convert");
+        }
+
+        String splitNumber[] = String.valueOf(doubleNumber).replace('.', '#').split("#");
+
+        // Descompone el trio de millones 
+        int millon = Integer.parseInt(String.valueOf(getDigitAt(splitNumber[0],
+                8))
+                + String.valueOf(getDigitAt(splitNumber[0], 7))
+                + String.valueOf(getDigitAt(splitNumber[0], 6)));
+        if (millon == 1) {
+            converted = "A million ";
+        }
+        if (millon > 1) {
+            converted = convertNumber(String.valueOf(millon)) + "Millions ";
+        }
+
+        // Descompone el trio de miles - 
+        int miles = Integer.parseInt(String.valueOf(getDigitAt(splitNumber[0],
+                5))
+                + String.valueOf(getDigitAt(splitNumber[0], 4))
+                + String.valueOf(getDigitAt(splitNumber[0], 3)));
+        if (miles == 1) {
+            converted += "One Thousand "; 
+        }
+        if (miles > 1) {
+            converted += convertNumber(String.valueOf(miles)) + "Thousand ";
+        }
+
+        // Descompone el ultimo trio de unidades
+        int cientos = Integer.parseInt(String.valueOf(getDigitAt(
+                splitNumber[0], 2))
+                + String.valueOf(getDigitAt(splitNumber[0], 1))
+                + String.valueOf(getDigitAt(splitNumber[0], 0)));
+        if (cientos == 1) {
+            converted += "One";
+        }
+
+        if (millon + miles + cientos == 0) {
+            converted += "Zero dollars";
```

Comment:
String does not allow appending. Each method you invoke on a String creates a new object and returns it. This is because String is immutable - it cannot change its internal state.

On the other hand StringBuilder is mutable. When you call append(..) it alters the internal char array, rather than creating a new string object.

Thus it is more efficient to have:

StringBuilder sb = new StringBuilder();
for (int i = 0; i < 500; i ++) {
    sb.append(i);
}
rather than str += i, which would create 500 new string objects.

Note that in the example I use a loop. As helios notes in the comments, the compiler automatically translates expressions like String d = a + b + c to something like

String d = new StringBuilder(a).append(b).append(c).toString();
Note also that there is StringBuffer in addition to StringBuilder. The difference is that the former has synchronized methods. If you use it as a local variable, use StringBuilder. If it happens that it's possible for it to be accessed by multiple threads, use StringBuffer (that's rarer)